tests/testlib/s2n_ktls_test_utils.c

```diff
@@ -0,0 +1,219 @@
+    uint8_t *buf = msg->msg_iov->iov_base;
+    POSIX_ENSURE_REF(buf);
+
+    /* There is no data available so return blocked */
+    if (!s2n_stuffer_data_available(&io_ctx->ancillary_buffer)) {
+        errno = EAGAIN;
+        return -1;
+    }
+
+    ssize_t total_read = 0;
+    /* updated as partial or multiple records are read */
+    size_t updated_requested_len = msg->msg_iov->iov_len;
+    /* track two record_types since it is possible to read multiple records of the same type */
+    *record_type = 0;
+    uint8_t next_record_type = 0;
+    while (*record_type == next_record_type) {
+        /* read record_type and number of bytes available in the next record */
+        POSIX_GUARD(s2n_stuffer_read_uint8(&io_ctx->ancillary_buffer, record_type));
+        uint16_t n_avail = 0;
+        POSIX_GUARD(s2n_stuffer_read_uint16(&io_ctx->ancillary_buffer, &n_avail));
+        POSIX_ENSURE_LTE(n_avail, s2n_stuffer_data_available(&io_ctx->data_buffer));
+
+        /* read minimul of requested_len and bytes_available */
+        size_t n_read = MIN(updated_requested_len, n_avail);
+        POSIX_ENSURE_GT(n_read, 0);
+
+        int ret = s2n_stuffer_read_bytes(&io_ctx->data_buffer, buf + total_read, n_read);
+        if (ret < 0) {
+            errno = EINVAL;
+            return -1;
+        }
+
+        updated_requested_len -= n_read;
+        POSIX_ENSURE_GTE(updated_requested_len, 0);
+        total_read += n_read;
+
+        /* Handle if we partially read a record */
+        ssize_t remaining_len = n_avail - n_read;
+        if (remaining_len) {
+            POSIX_GUARD_RESULT(s2n_test_ktls_update_prev_header_len(io_ctx, remaining_len));
+        }
+
+        /* if already read the requested amount then break */
+        if (updated_requested_len == 0) {
+            break;
+        }
+        /* Attempt to read multiple records (must be of the same type) */
+        if (updated_requested_len) {
+            int ret = s2n_stuffer_peek_char(&io_ctx->ancillary_buffer, (char *) &next_record_type);
+
+            bool no_more_records = ret < 0;
+            bool next_record_different_type = next_record_type != *record_type;
+            if (no_more_records || next_record_different_type) {
+                break;
+            }
+        }
+    }
+
+    return total_read;
+}
+
+S2N_RESULT s2n_test_validate_data(struct s2n_test_ktls_io_stuffer *ktls_io, uint8_t *expected_data, uint16_t expected_len)
+{
+    RESULT_ENSURE_REF(ktls_io);
+    RESULT_ENSURE_REF(expected_data);
+
+    struct s2n_stuffer *stuffer = &ktls_io->data_buffer;
+    RESULT_ENSURE_EQ(s2n_stuffer_data_available(stuffer), expected_len);
+    uint8_t *data_ptr = s2n_stuffer_raw_read(stuffer, expected_len);
+    RESULT_ENSURE_REF(data_ptr);
+    RESULT_ENSURE_EQ(memcmp(data_ptr, expected_data, expected_len), 0);
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_test_validate_ancillary(struct s2n_test_ktls_io_stuffer *ktls_io, uint8_t expected_record_type, uint16_t expected_len)
```

Comment:
I think it would make more sense if `s2n_test_validate_ancillary`, `s2n_test_validate_data` , `s2n_test_init_ktls_stuffer_io`, and `s2n_ktls_io_pair_free` were all static methods in the ktls test file. My reasoning is that they're all functions that are kind of made to make writing your test easier. They make more sense in the context of that test.

However if you're going to be writing more test files in the future you would have to move them out.

Reply:
yep the idea is to reuse them in 2 other upcoming PRs otherwise yea they would belong in the test itself